algo/numeric/simplex.h

```diff
@@ -0,0 +1,201 @@
+#pragma once
+
+#include "numeric/matrix.h"
+
+#include <cassert>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <vector>
+
+namespace algo {
+
+// Solves linear programming tasks in the following form:
+// fs[0] * xs[0] + fs[1] * xs[1] + ... + fs[n - 1] * xs[n - 1] -> max
+//
+// as[0][0] * xs[0] + as[0][1] * xs[1] + ... + as[0][n - 1] * xs[n - 1] <= bs[0]
+// as[1][0] * xs[0] + as[1][1] * xs[1] + ... + as[1][n - 1] * xs[n - 1] <= bs[1]
+// ...
+// as[m - 1][0] * xs[0] + as[m - 1][1] * xs[1] + ... + as[m - 1][n - 1] * xs[n - 1] <= bs[m - 1]
+//
+// Where for all i: bs[i] >= 0, xs[i] >= 0.
+//
+// Inspired by little-known and hard-to-find, but very nice paper by
+// D.E.K.: https://www-cs-faculty.stanford.edu/~knuth/programs/lp.w
+class Simplex {
+public:
+  static inline const double kEps = 1e-10;
+
+  template <typename M>
+  explicit Simplex(const M& as, const std::vector<double>& bs, const std::vector<double>& fs, bool verbose = false)
+      : m_as(as.Height() + 1, as.Width() + as.Height() + 1)
+      , m_trial(m_as.Width())
+      , m_brow(m_as.Width())
+      , m_bcol(m_as.Height())
+      , m_verbose(verbose) {
+    assert(as.Height() == bs.size());
+    assert(as.Width() == fs.size());
+
+    for (size_t i = 0; i < fs.size(); ++i)
```

Comment:
nit: A convention that I find helpful is to use ```j``` to index columns even in a larger scope. If you write ```m_as(0, as.Height() + 1 + j) = -fs[j];``` here, this line will be easier to compare to with Line 48.

Reply:
Done